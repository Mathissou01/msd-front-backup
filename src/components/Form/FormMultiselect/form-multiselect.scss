@use "src/styles/01-settings/settings.colors" as colors;

.c-FormMultiselect {
  display: flex;
  flex-direction: column;
  gap: 24px;
  width: 100%;
}
