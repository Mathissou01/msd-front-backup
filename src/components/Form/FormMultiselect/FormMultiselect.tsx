import { useFormContext } from "react-hook-form";
import { ErrorMessage } from "@hookform/error-message";
import React, { useEffect, useMemo, useRef, useState } from "react";
import { isString, removeNulls } from "../../../lib/utilities";
import CommonFormErrorText from "../../Common/CommonFormErrorText/CommonFormErrorText";
import FormLabel from "../FormLabel/FormLabel";
import "./form-multiselect.scss";

export interface IOptionWrapper<T> {
  group?: string;
  label?: string;
  option: T | null;
}

export function mapOptionsInWrappers<T extends { [index: string]: unknown }>(
  options: Array<T | null>,
  fieldToGroup?: string,
): Array<IOptionWrapper<T>> {
  const mappedOptions: Array<IOptionWrapper<T> | null> = options.map(
    (option) => {
      const group = option && fieldToGroup ? option[fieldToGroup] : null;
      return option
        ? {
            option: option,
            group: isString(group) ? group : undefined,
          }
        : null;
    },
  );
  return mappedOptions.filter(removeNulls);
}

interface IFormMultiselectProps<T> {
  name: string;
  label: string;
  secondaryLabel?: string;
  displayTransform: (...args: Array<T>) => string;
  isRequired?: boolean;
  isDisabled?: boolean;
  selectAmount: number;
  options: Array<IOptionWrapper<T>>;
  optionKey: keyof T & string;
  defaultValues?: Array<T>;
  noneSelectedLabel?: string;
}

export default function FormMultiselect<T>({
  name,
  label,
  secondaryLabel,
  displayTransform,
  isRequired = false,
  isDisabled = false,
  options,
  selectAmount,
  optionKey,
  defaultValues,
  noneSelectedLabel,
}: IFormMultiselectProps<T>) {
  /* Static Data */
  noneSelectedLabel = noneSelectedLabel ?? "- Sélectionnez un élément -";
  const errorMessages = {
    required: "Ce champ est obligatoire",
  };

  /* Local Data */
  const [selectedIndexes, setSelectedIndexes] = useState<Array<number>>(
    Array(selectAmount).fill(-1),
  );
  const indexesRef = useRef(selectedIndexes);
  const {
    register,
    watch,
    setValue,
    formState: { isSubmitting, errors },
  } = useFormContext();
  for (let i = 0; i < selectAmount; i++) {
    register(`${name}_${i}`, {
      required: {
        value: isRequired,
        message: errorMessages.required,
      },
      value: defaultValues?.[i] ?? null,
    });
  }
  const watchValues: Array<Record<string, unknown>> = useMemo(() => {
    const values = [];
    for (let i = 0; i < selectAmount; i++) {
      values[i] = watch(`${name}_${i}`);
    }
    return values;
  }, [name, selectAmount, watch]);

  useEffect(() => {
    indexesRef.current = selectedIndexes;
  }, [selectedIndexes]);

  useEffect(() => {
    const updatedArray = [...indexesRef.current];
    for (let i = 0; i < selectAmount; i++) {
      updatedArray[i] = options.findIndex(
        (wrapper) =>
          wrapper?.option &&
          wrapper.option?.[optionKey] === watchValues[i]?.[optionKey],
      );
    }
    setSelectedIndexes(updatedArray);
  }, [selectAmount, options, optionKey, watchValues]);

  const renderGroups = (i: number) => {
    let persistentIndex = 0;
    return [...new Map(options.map((item) => [item["group"], item]))].map(
      (group, index) => {
        const initialIndex = persistentIndex;
        const groupOptions = options.filter(
          (option) => option.group === group[0],
        );
        persistentIndex = persistentIndex + groupOptions.length;
        return (
          groupOptions.length > 0 && (
            <optgroup
              className="o-SelectWrapper__OptGroup"
              key={`${name}_${i}_${index}_${group[0]}`}
              label={group[0]}
            >
              {renderOptions(i, initialIndex, groupOptions)}
            </optgroup>
          )
        );
      },
    );
  };

  const renderOptions = (
    i: number,
    initialIndex: number,
    options: Array<IOptionWrapper<T>>,
  ) => {
    return options.map((wrapper, index) => {
      const totalIndex = initialIndex + index;
      return (
        wrapper &&
        (totalIndex === selectedIndexes[i] ||
          !selectedIndexes.includes(totalIndex)) && (
          <option
            className="o-SelectWrapper__Option"
            key={`${name}_${i}_${totalIndex}`}
            value={totalIndex}
          >
            {wrapper.option && !wrapper.label
              ? displayTransform(wrapper.option)
              : wrapper.label}
          </option>
        )
      );
    });
  };

  return (
    <div className="c-FormMultiselect">
      {Array.from(Array(selectAmount), (_, i) => (
        <div key={`${name}_${i}`}>
          <FormLabel
            forId={`${name}_${i}`}
            label={`${label} n°${i + 1}`}
            isRequired={isRequired}
            secondaryLabel={secondaryLabel}
          />
          <div className="o-SelectWrapper">
            <select
              className={`o-SelectWrapper__Select ${
                selectedIndexes[i] < 0
                  ? "o-SelectWrapper__Select_placeholder"
                  : ""
              } ${
                errors[`${name}_${i}`] ? "o-SelectWrapper__Select_invalid" : ""
              }`}
              id={`${name}_${i}`}
              value={selectedIndexes[i]}
              onChange={(event) => {
                const index = Number.parseInt(event.target.value);
                const updatedArray = [...selectedIndexes];
                updatedArray[i] = index;
                setSelectedIndexes(updatedArray);
                setValue(`${name}_${i}`, options[index]?.option ?? null, {
                  shouldValidate: true,
                  shouldDirty: true,
                });
              }}
              disabled={isSubmitting || isDisabled}
              aria-invalid={!!errors[`${name}_${i}`]}
              aria-errormessage={`${name}_${i}_error`}
              data-testid={`form-multiselect_${i}`}
            >
              <option disabled={isRequired} hidden={isRequired} value={-1}>
                {noneSelectedLabel}
              </option>
              {options.every((wrapper) => wrapper.group)
                ? renderGroups(i)
                : renderOptions(i, 0, options)}
            </select>
          </div>
          <ErrorMessage
            errors={errors}
            name={`${name}_${i}`}
            render={({ message }: { message: string }) => (
              <CommonFormErrorText
                message={message}
                errorId={`${name}_${i}_error`}
              />
            )}
          />
        </div>
      ))}
    </div>
  );
}
