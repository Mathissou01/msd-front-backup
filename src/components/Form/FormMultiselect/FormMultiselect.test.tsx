import { render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import React, { ReactNode } from "react";
import FormMultiselect from "./FormMultiselect";

const mock = {
  name: "form-multiselect",
  label: "form multiselect",
};

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return (
    <>
      <FormProvider {...formMethods}>{props.children}</FormProvider>
    </>
  );
};

describe("FormMultiselect", () => {
  it("renders", () => {
    const handleTransform = (data: { id: number; content: string }) => {
      return "-" + data.content;
    };
    const { container } = render(
      <Wrapper>
        <FormMultiselect
          name={mock.name}
          label={mock.label}
          displayTransform={handleTransform}
          isRequired={false}
          isDisabled={false}
          selectAmount={2}
          options={[
            { option: { id: 1, content: "content 1" } },
            { option: { id: 2, content: "content 2" } },
          ]}
          optionKey={"id"}
        />
      </Wrapper>,
    );
    expect(screen.getByTestId("form-multiselect_0")).toBeInTheDocument;
    expect(screen.getByTestId("form-multiselect_1")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
