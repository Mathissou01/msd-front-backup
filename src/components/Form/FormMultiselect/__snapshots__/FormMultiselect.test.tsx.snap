// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormMultiselect renders 1`] = `
<div>
  <div
    class="c-FormMultiselect"
  >
    <div>
      <div
        class="c-FormLabel"
      >
        <label
          class="c-FormLabel__LabelWrapper"
          for="form-multiselect_0"
        >
          <span
            class="c-FormLabel__Label"
          >
            form multiselect n°1
          </span>
        </label>
      </div>
      <div
        class="o-SelectWrapper"
      >
        <select
          aria-errormessage="form-multiselect_0_error"
          aria-invalid="false"
          class="o-SelectWrapper__Select o-SelectWrapper__Select_placeholder "
          data-testid="form-multiselect_0"
          id="form-multiselect_0"
        >
          <option
            value="-1"
          >
            - Sélectionnez un élément -
          </option>
          <option
            class="o-SelectWrapper__Option"
            value="0"
          >
            -content 1
          </option>
          <option
            class="o-SelectWrapper__Option"
            value="1"
          >
            -content 2
          </option>
        </select>
      </div>
    </div>
    <div>
      <div
        class="c-FormLabel"
      >
        <label
          class="c-FormLabel__LabelWrapper"
          for="form-multiselect_1"
        >
          <span
            class="c-FormLabel__Label"
          >
            form multiselect n°2
          </span>
        </label>
      </div>
      <div
        class="o-SelectWrapper"
      >
        <select
          aria-errormessage="form-multiselect_1_error"
          aria-invalid="false"
          class="o-SelectWrapper__Select o-SelectWrapper__Select_placeholder "
          data-testid="form-multiselect_1"
          id="form-multiselect_1"
        >
          <option
            value="-1"
          >
            - Sélectionnez un élément -
          </option>
          <option
            class="o-SelectWrapper__Option"
            value="0"
          >
            -content 1
          </option>
          <option
            class="o-SelectWrapper__Option"
            value="1"
          >
            -content 2
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;
